### ISSUE TITLE
ConvTranspose2d much slower with output_padding set

### ISSUE THREAD

**Issue by zplizzi** [pytorch/pytorch#31690]
## 🐛 Bug

A `nn.ConvTranspose2d` can be up to two orders of magnitude slower when the `output_padding` option is set to a nonzero value then when it isn't.

## To Reproduce

```
class Test(nn.Module):
    def __init__(self, output_padding):
        super().__init__()
        
        self.c1 = nn.ConvTranspose2d(128, 64, kernel_size=3, stride=2,
                                     padding=1, output_padding=output_padding)
        self.c2 = nn.ConvTranspose2d(64, 32, kernel_size=3, stride=2,
                                     padding=1, output_padding=output_padding)
        self.c3 = nn.ConvTranspose2d(32, 3, kernel_size=3, stride=2,
                                     padding=1, output_padding=output_padding)
    def forward(self, x):
        x = self.c1(x)
        x = nn.functional.relu(x)
        x = self.c2(x)
        x = nn.functional.relu(x)
        x = self.c3(x)
        x = nn.functional.relu(x)
        return x

def test(output_padding):
    network = Test(output_padding=output_padding).cuda()
    x = torch.rand((1028, 128, 6, 6)).cuda()


    for i in range(5):
        start = time.time()
        for i in range(10):
            network(x)
        print((time.time() - start) / 10)

test((0, 0))
test((1, 1))
```

On my V100 machine, I get timings of about 0.0004s per forward pass with `output_padding=(0, 0)` and about .05s with `output_padding=(1, 1)`.

## Expected behavior

Given that the `output_padding` option has minimal effect on the number of FLOPs required to compute the function, I would expect the computation time to be similar with and without the option.

## Environment

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 16.04.6 LTS
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 418.87.00
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.2

Versions of relevant libraries:
[pip3] numpy==1.17.4
[pip3] torch==1.3.1
[pip3] torchcule==0.1.0
[pip3] torchvision==0.4.1
[conda] Could not collect

cc @csarofeen @ptrblck

**Comment by ngimel** [comment]
This likely happens because cudnn is not used for convolutions with non-zero `output_padding`, which in turn is a result of this check 
https://github.com/pytorch/pytorch/blob/22d84204f7188ed4042505e723eef46f048ff47c/aten/src/ATen/native/Convolution.cpp#L96-L102
This check was added because wrong results/CUDNN_STATUS_BAD_PARAMS were produced in some similar cases, but maybe the check is unnecessarily strict. 
Looks like `output_padding[i] >= dilation[i]` will be true for most cases. T
@ptrblck, @csarofeen to check if it is still a real limitation for cudnn, or if this check can be relaxed.

**Comment by linjiapro** [comment]
@xwang233 @ngimel  thanks for investigating and fixing this. I am experiencing exactly the same issue. Do we have a plan to push the fix to 1.7 release branch?

**Comment by linjiapro** [comment]
This is my temporary workaround for this issue. 
There is a formula about how to calculate output size:
https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html

In real usage cases, the output size and input size are usually fixed, and we need to tune the padding, dilation, etc. We can set output_padding = 0, and design other parameters to make sure output size and input size is fit. In my case, I changed kernel size to match input/.output size while having output_padding = 0.
